### DISCUSSION TITLE
Request to add a set of in-page navigation components/hooks

### DISCUSSION THREAD

**Discussion by mumingpo** [mantinedev/mantine#2152] [Feature requests]
This is a feature request for a commonly used web design pattern by using hash components in a URL to navigate within a page. I sort of have it implemented already in Next.js but I haven't rigorously tested it yet. I can provide the source code of my implementation and help implement it if needed.

The change is to:

1. Add a `NavbarContext` provider (can be merged into other existing context providers);
2. Add a `PageNavLinks` component;
3. Add a `PageAnchor` component;
4. Add a `useInPageNavigation` hook;

## NavbarContext:

This provides a context to store the `ref` objects generated by the `PageAnchors` in a `React.useState`.

- `PageNavLinks` read this to generate nested `NavLink`s for the `Navbar` component.
- `PageAnchor` updates this to add/remove `ref` objects by using the `React.useEffect` hook.
- `useInPageNavigation` hook reads the `ref` objects to `ref.current.scrollIntoView()`.

### The PageAnchors object

The tree-like data structure used to store the `ref` objects inside `NavbarContext`.
This is initialized with `{ '_root', { ref: null, label: 'On this page', children: [] }` to represent the root element of the document.
This value and a setter method is provided by the `NavbarContext`.

```typescript
type PageAnchors = Record<string, {
  ref: null | React.MutableRefObject<HTMLDivElement | null>,
  label: string,
  children: Array<string>,
  icon?: JSX.Element,
}>;
```

## PageNavLinks

This reads the stored anchors from `_root` and recursively generate `NavLink`s with `href={'#' + key}` according to the structure of the `PageAnchors` object.

- This reads `NavbarContext`

## PageAnchor

This is a `<div />` component with a height of 0 inserted into the DOM. It inserts/removes ref objects into the `PageAnchors` by using the `React.useEffect` hook.

- This updates `NavbarContext`

## useInPageNavigation

This is a hook that reads the `#location` from window.location/router/whatever, and uses the respective ref object to perform `ref.current.scrollIntoView()`.

- This reads `NavbarContext`

## Demo

Code developers would write:

![NavbarContext](https://user-images.githubusercontent.com/25801035/185272183-863c84c4-ced7-48e0-8f4e-58076cfc88fe.png)
![PageNavLinks](https://user-images.githubusercontent.com/25801035/185272207-4c3f9a38-6ff1-48d4-a0d7-3b960f13018a.png)
![usePageNav_and_PageAnchor](https://user-images.githubusercontent.com/25801035/185272231-b8fa2853-5443-4944-8462-5469b7d09634.png)

How it would appear in a webpage:

![effect](https://user-images.githubusercontent.com/25801035/185272490-b02cad19-23d9-4f8e-ab25-56830b6a9ad8.png)

**Comment by mumingpo** [comment]
My current implementation:
[bundle.zip](https://github.com/mantinedev/mantine/files/9366947/bundle.zip)

**Comment by rtivital** [comment]
Hi @mumingpo, thanks for a feature request. This is outside of UI library scope – we should not add this.

**Reply by mumingpo** [reply]
Ok!
